clients/drcachesim/tests/cache_replacement_policy_unit_test.cpp

```diff
@@ -409,6 +410,140 @@ unit_test_cache_lfu_eight_way()
+    cache_rrip_test.access_and_check(addr_vec[ADDR_A], 2);     // h : A0 C1 d2 E2
+    cache_rrip_test.access_and_check(addr_vec[ADDR_B], 3);     // mp: A1 C2 B2 e3
+    cache_rrip_test.access_and_check(addr_vec[ADDR_C], 3);     // h : A1 C0 B2 e3
+    cache_rrip_test.access_and_check(addr_vec[ADDR_D], 2);     // m : A1 C0 b2 D2
+    cache_rrip_test.access_and_check(addr_vec[ADDR_E], 3);     // mp: A2 C1 E2 d3
+}
+
+void
+unit_test_cache_brrip_four_way()
+{
+    caching_device_policy_test_t<cache_t> cache_rrip_test(/*associativity=*/4,
+                                                          /*line_size=*/32);
+    cache_rrip_test.initialize_cache(std::unique_ptr<policy_rrip_t>(new policy_rrip_t(
+                                         /*num_blocks=*/256 / 4, /*associativity=*/4,
+                                         /*rrpv_bits=*/2, /*rrpv_period=*/4,
+                                         /*rrpv_long_per_period=*/1)),
+                                     256);
+
+    assert(cache_rrip_test.get_replace_policy() == "RRIP");
+    assert(cache_rrip_test.block_indices_are_identical(addr_vec));
+    assert(cache_rrip_test.tags_are_different(addr_vec));
+
+    // m: miss
+    // h: hit
+    // p: promotion
+    // l: "long" RRPV == 2
+    // d: "distant" RRPV == 3
+
+    // Cache initialization
+    cache_rrip_test.access_and_check(addr_vec[ADDR_A], 1); //     ml : A2 x3 X3 X3
+    cache_rrip_test.access_and_check(addr_vec[ADDR_B], 2); //     md : A2 B3 x3 X3
+    cache_rrip_test.access_and_check(addr_vec[ADDR_C], 3); //     md : A2 B3 C3 x3
+    cache_rrip_test.access_and_check(addr_vec[ADDR_D], 1); //     md : A2 b3 C3 D3
+    cache_rrip_test.access_and_check(addr_vec[ADDR_A], 1); //     h  : A0 b3 C3 D3
+    cache_rrip_test.access_and_check(addr_vec[ADDR_D], 1); //     h  : A0 b3 C3 D0
+    cache_rrip_test.access_and_check(addr_vec[ADDR_B], 2); //     h  : A0 B0 c3 D0
+    cache_rrip_test.access_and_check(addr_vec[ADDR_C], 0); //     h  : a0 B0 C0 D0
+
+    // Cache pollution: 1 block reused, 2 cached
+    cache_rrip_test.invalidate_and_check(addr_vec[ADDR_D], 3); //      A0 B0 C0 x3
+    cache_rrip_test.invalidate_and_check(addr_vec[ADDR_C], 2); //      A0 B0 x3 X3
+    cache_rrip_test.invalidate_and_check(addr_vec[ADDR_B], 1); //      A0 x3 X3 X3
+    cache_rrip_test.access_and_check(addr_vec[ADDR_A], 1);     // h  : A0 x3 X3 X3
+    cache_rrip_test.access_and_check(addr_vec[ADDR_B], 2);     // ml : A0 B2 x3 X3
+    cache_rrip_test.access_and_check(addr_vec[ADDR_C], 3);     // md : A0 B2 C3 x3
+    cache_rrip_test.access_and_check(addr_vec[ADDR_D], 2);     // md : A0 B2 c3 D3
+    cache_rrip_test.access_and_check(addr_vec[ADDR_E], 2);     // md : A0 B2 e3 D3
+    cache_rrip_test.access_and_check(addr_vec[ADDR_A], 2);     // h  : A0 B2 e3 D3
+    cache_rrip_test.access_and_check(addr_vec[ADDR_B], 2);     // h  : A0 B0 e3 D3
+    cache_rrip_test.access_and_check(addr_vec[ADDR_C], 3);     // ml : A0 B0 C2 d3
```

Comment:
C was added with "long" rrpv, but I don't see how the subsequent accesses in this test case help prove that. It would help if we had an access to some new line "F" which would replace "d" first (since it was added with the "distant" rrpv), and only then the next access to "E" would replace "c".